File: src/sampling.ts
import path from "node:path";

import { AdapterError } from "./adapters/index.js";
import type { Adapter, WarnSink } from "./adapters/types.js";
import type { ConfigKinds, ConfigMetrics, ResolvedMetricMeta } from "./config.js";
import { resolveMetricMeta } from "./config.js";
import { GymratError, messageOf } from "./errors.js";
import { exec } from "./exec.js";
import type { ExecResult, ExecTimeoutError } from "./exec.js";
import { computeHalfRange, computeMedian } from "./math.js";
import { formatCleanupFailures } from "./report/text.js";
import { installTerminationCleanup } from "./signals.js";
import { cleanupWorktrees, planWorktree, materializeWorktree } from "./targets.js";
import type { CleanupResult, Target, WorktreeInfo } from "./targets.js";

/** A prepare step about to run for a target with a prepare script. */
interface PrepareProgressStep {
  kind: "prepare";
  label: string;
}

/** A bench/sample step about to run — 1-based index within the total sample count. */
interface SampleProgressStep {
  kind: "sample";
  index: number;
  total: number;
  label: string;
}

/** Structured step info emitted at the start of each prepare or sample step. */
export type ProgressStep = PrepareProgressStep | SampleProgressStep;

// ---------------------------------------------------------------------------
// Command-error formatting
// ---------------------------------------------------------------------------

/** Fields that identify which command failed and where, for structured error reporting. */
export interface CommandErrorContext {
  phase: "prepare" | "bench";
  position?: "old" | "new";
  label: string;
  command: string;
  target: Target;
  dir: string;
  sample?: number;
}

/** The header line naming the command, its verb, and where it ran. */
function formatCommandErrorHeader(context: CommandErrorContext, isTimeout: boolean): string {
  const verb = isTimeout ? "timed out" : "failed";
  const positionPart = context.position !== undefined ? `${context.position}, ` : "";
  const samplePart = context.sample !== undefined ? `, sample ${context.sample}` : "";
  return `${context.phase} command ${verb} (${positionPart}"${context.label}"${samplePart})`;
}

/** The lines naming where the command ran: a worktree and its ref, or a plain directory. */
function formatCommandErrorLocation(context: CommandErrorContext): string[] {
  return context.target.kind === "ref"
    ? [`  ref:       ${context.target.ref}`, `  worktree:  ${context.dir}`]
    : [`  dir:       ${context.dir}`];
}

/**
 * The captured stdout and stderr, labeled when both are present so a reader can
 * tell them apart, unlabeled when only one stream has anything to show.
 */
function formatStreamEntry(label: string, text: string, totalBytes: number): string[] {
  const capturedBytes = Buffer.byteLength(text, "utf8");
  const suffix = totalBytes > capturedBytes ? ` (truncated, ${totalBytes} bytes total)` : "";
  return [`--- ${label}${suffix} ---`, text];
}

function formatCommandErrorOutput(failure: ExecResult | ExecTimeoutError): string[] {
  const streams = (
    [
      ["stderr", failure.stderr, failure.stderrBytes],
      ["stdout", failure.stdout, failure.stdoutBytes],
    ] satisfies [string, string, number][]
  ).filter(([, text]) => text.length > 0);

  if (streams.length < 2) {
    const entry = streams[0];
    if (entry === undefined) return [];
    const [, text, totalBytes] = entry;
    const capturedBytes = Buffer.byteLength(text, "utf8");
    return totalBytes > capturedBytes ? formatStreamEntry(entry[0], text, totalBytes) : [text];
  }
  return streams.flatMap(([label, text, totalBytes]) => formatStreamEntry(label, text, totalBytes));
}

function formatCommandError(
  context: CommandErrorContext,
  failure: ExecResult | ExecTimeoutError,
): string {
  const isTimeout = "kind" in failure;

  return [
    formatCommandErrorHeader(context, isTimeout),
    ...formatCommandErrorLocation(context),
    `  command:   ${context.command}`,
    isTimeout ? `  timeout:   ${failure.timeoutMs}ms` : `  exit code: ${failure.exitCode}`,
    ...formatCommandErrorOutput(failure),
  ].join("\n");
}

/**
 * Structured error for a command that failed during a benchmark or prepare phase.
 *
 * The formatted message carries the full context: phase, position, target, and the
 * exit code or timeout. Ref-target failures append a hint about the ref possibly
 * lacking the files the command needs.
 */
export class CommandError extends GymratError {
  constructor(context: CommandErrorContext, failure: ExecResult | ExecTimeoutError) {
    const hint =
      context.target.kind === "ref"
        ? "the worktree only contains files tracked at this ref; untracked, gitignored, or not-yet-committed files are absent"
        : undefined;
    super(formatCommandError(context, failure), hint);
  }
}

// ---------------------------------------------------------------------------
// Sampling core
// ---------------------------------------------------------------------------

/**
 * One target of a comparison or measurement run.
 *
 * `target` is either a git ref (resolved to a throwaway worktree) or a
 * filesystem directory path (benched in place).
 */
export interface TargetSpec {
  target: string;
  /** Display label; defaults to the ref name or the directory's base name. */
  label?: string;
}

/** A target resolved to its working directory, label, and optional role marker. */
export interface TargetContext {
  target: Target;
  dir: string;
  label: string;
  position?: "old" | "new";
}

/** One target's measurements, kept beside the target that produced them. */
export interface TargetSamples {
  readonly ctx: TargetContext;
  readonly samples: Record<string, number>[];
}

/**
 * The caller-facing subset of run configuration that `collectSamples` reads.
 *
 * `onProgress` and `warn` are optional sinks: omitted, progress goes
 * unreported and the adapter's own default (stderr) takes warnings.
 */
export interface SamplingOptions {
  /** Run through the shell in each target's directory. */
  bench: string;
  prepare?: string;
  /** How many rounds to run; each round runs `bench` once per target. */
  samples: number;
  timeoutSeconds: number;
  /** Fire-and-forget callback invoked at the start of each prepare or sample step. */
  onProgress?: (step: ProgressStep) => void;
  /**
   * Where the adapter's complaints about unreadable bench output go. Omitted,
   * the adapter falls back to stderr.
   */
  warn?: WarnSink;
}

/**
 * The run settings a comparison and a measurement both take, beyond the targets
 * each names for itself.
 *
 * `SamplingOptions` is what `collectSamples` reads; the three fields added here
 * are what the caller needs to turn raw samples into a report — which parser to
 * read the bench output with, and the per-metric and per-kind overrides that
 * settle each metric's metadata.
 */
export interface RunOptions extends SamplingOptions {
  /** Which output format `bench` writes: `"metric-lines"` or `"mitata"`. */
  adapter: string;
  configMetrics?: ConfigMetrics;
  configKinds?: ConfigKinds;
}

/**
 * Run a shell command and throw on timeout or non-zero exit.
 *
 * Aborting `signal` kills the command's whole process group; bench commands are
 * spawned detached, so a Ctrl-C delivered to gymrat never reaches them by itself.
 */
async function runCommand(
  phase: "prepare" | "bench",
  ctx: TargetContext,
  command: string,
  timeoutMs: number,
  signal: AbortSignal,
  sample?: number,
): Promise<string> {
  const context: CommandErrorContext = { ...ctx, phase, command, sample };

  const result = await exec(command, { cwd: ctx.dir, timeoutMs, signal });

  if ("kind" in result || result.exitCode !== 0) {
    throw new CommandError(context, result);
  }

  return result.stdout;
}

/**
 * Collect samples round-robin across the targets in the order given.
 *
 * Each round runs the bench once per target, in the given order, before the next
 * round starts. `prepare` runs once per target, up front.
 *
 * The caller decides the ordering and assigns roles — the sampling core treats
 * every target identically.
 */
export async function collectSamples(
  adapter: Adapter,
  targets: readonly TargetContext[],
  options: SamplingOptions,
  signal: AbortSignal,
): Promise<TargetSamples[]> {
  const timeoutMs = options.timeoutSeconds * 1000;
  const collected: TargetSamples[] = targets.map((ctx) => ({ ctx, samples: [] }));

  if (options.prepare) {
    for (const { ctx } of collected) {
      options.onProgress?.({ kind: "prepare", label: ctx.label });
      await runCommand("prepare", ctx, options.prepare, timeoutMs, signal);
    }
  }

  for (let round = 0; round < options.samples; round++) {
    for (const { ctx, samples } of collected) {
      options.onProgress?.({
        kind: "sample",
        index: round + 1,
        total: options.samples,
        label: ctx.label,
      });
      const stdout = await runCommand("bench", ctx, options.bench, timeoutMs, signal, round + 1);
      samples.push(adapter.parse(stdout, options.warn));
    }
  }

  return collected;
}

// ---------------------------------------------------------------------------
// Target resolution
// ---------------------------------------------------------------------------

/**
 * Resolve the working directory for a target, creating a worktree for ref targets.
 *
 * The worktree is registered before `git worktree add` runs, not after: git can
 * be killed once the directory is on disk but before the command returns, and
 * every cleanup path — the failure path and the termination handler alike —
 * sweeps only what `worktrees` already names.
 */
export function resolveDir(resolved: Target, repoDir: string, worktrees: WorktreeInfo[]): string {
  if (resolved.kind === "ref") {
    const worktree = planWorktree(resolved);
    worktrees.push(worktree);
    materializeWorktree(worktree, repoDir);
    return worktree.dir;
  }
  return resolved.dir;
}

/**
 * The target's display label: the explicit label if one was given, otherwise
 * the ref name or the directory's base name.
 */
export function resolveLabel(explicit: string | undefined, resolved: Target): string {
  return explicit ?? (resolved.kind === "ref" ? resolved.ref : path.basename(resolved.dir));
}

// ---------------------------------------------------------------------------
// Metric statistics
// ---------------------------------------------------------------------------

/**
 * Half the observed range as a percentage of the median — the run-to-run jitter a
 * verdict's noise band is judged against.
 *
 * Half-range rather than full range so the figure reads as "± this much" either
 * side of the median, which is how the report prints it. A zero median has no
 * scale to be a percentage of, so it contributes no spread at all.
 *
 * A single observation has no run-to-run jitter to report: its range is zero
 * only because nothing was ever compared against it, and `± 0%` would state
 * that as a measured result. Such a side has no spread at all.
 */
function computeSpread(values: readonly number[], median: number): number | undefined {
  if (values.length < 2 || median === 0) return undefined;
  const ratio = (computeHalfRange(values) / Math.abs(median)) * 100;
  return Number.isFinite(ratio) ? ratio : undefined;
}

/** A side's median and spread over the given values, or both undefined when there are none. */
export function computeMetricStats(values: readonly number[]): {
  median?: number;
  spread?: number;
} {
  if (values.length === 0) {
    return {};
  }
  const median = computeMedian(values);
  return { median, spread: computeSpread(values, median) };
}

/** Every value a side reported for a metric, regardless of what the other side has. */
export function ownValues(
  samples: readonly Record<string, number>[],
  metricName: string,
): number[] {
  return samples.map((sample) => sample[metricName]).filter((v) => v !== undefined);
}

/**
 * The values one side's displayed median is read from: the rounds the metric
 * paired in, or — when nothing paired — every round that side reported it.
 *
 * A metric that paired has a verdict its median must stay consistent with, so
 * the median comes from the paired rounds alone. A metric only one side ever
 * reported has no verdict to agree with, so falling back to its own rounds
 * shows a real measurement instead of an empty cell.
 */
export function pairedOrOwnValues(
  paired: readonly number[],
  samples: readonly Record<string, number>[],
  metricName: string,
): readonly number[] {
  return paired.length > 0 ? paired : ownValues(samples, metricName);
}

/** Every metric name any target reported, so a one-sided metric still gets a row. */
function collectMetricNames(sampleSets: readonly Record<string, number>[][]): Set<string> {
  return new Set(sampleSets.flat().flatMap(Object.keys));
}

/**
 * Collect metric names from sample sets and resolve their metadata in one step.
 *
 * Every call site that measures — `measure`, `compare`, `iterate` — runs the
 * same sequence: collect names, guard against an empty set, resolve metadata.
 * The `v8 ignore` guard covers the empty-set branch that adapters' own
 * validation makes unreachable in practice.
 */
export function resolveMetricMetaFromSamples(
  sampleSets: readonly Record<string, number>[][],
  configMetrics: ConfigMetrics | undefined,
  adapter: Adapter,
  configKinds?: ConfigKinds,
): Record<string, ResolvedMetricMeta> {
  const metricNames = collectMetricNames(sampleSets);

  /* v8 ignore if -- defensive check; adapters throw AdapterError for no metrics */
  if (metricNames.size === 0) {
    throw new GymratError("No metrics found in benchmark output");
  }

  return resolveMetricMeta(Array.from(metricNames), configMetrics, adapter, configKinds);
}

// ---------------------------------------------------------------------------
// Worktree orchestration
// ---------------------------------------------------------------------------

/**
 * Restate a failed run's error so it also names what cleanup left on disk.
 *
 * `cleanupWorktrees` reports rather than throws, so on the failure path the
 * original exception would otherwise carry the run's error and nothing else —
 * the caller prints that, and the user never learns which directories survived.
 * When cleanup was clean there is nothing to add and the original error is
 * returned untouched. The original always becomes the `cause`, keeping its
 * stack reachable.
 */
function withCleanupFailures(
  error: unknown,
  cleanup: { failures: readonly { dir: string; error: string }[]; pruneError: string | undefined },
): unknown {
  const details = formatCleanupFailures(cleanup.failures, cleanup.pruneError);

  if (details.length === 0) {
    return error;
  }

  const combinedMessage = [messageOf(error), "", "cleanup did not finish:", ...details].join("\n");

  const hint = error instanceof GymratError ? error.hint : undefined;

  return error instanceof AdapterError
    ? new AdapterError(combinedMessage, hint, { cause: error })
    : new GymratError(combinedMessage, hint, { cause: error });
}

/**
 * Run a target-resolution phase under the worktree/signal discipline `measure()`
 * and `compare()` both need, then build the caller's result from whatever the
 * phase produced.
 *
 * The result is built after `phase` settles so it states the cleanup that
 * actually ran, and cleanup runs exactly once per path — a second sweep could
 * succeed on a worktree the first one recorded as left behind, handing the
 * caller a report the disk contradicts. When `phase` throws, `buildResult` is
 * never called and the cleanup outcome rides out on the propagating error
 * instead.
 *
 * A failure raised later, inside `buildResult` itself, is not carried that
 * way: cleanup has already run by then and its outcome is dropped. That path
 * is defensive-only today, so it is left uncovered rather than guarded.
 *
 * SIGINT and SIGTERM take a third path: the in-flight command is killed,
 * cleanup is attempted, and the process exits with `128 + signum` without a
 * result. The handlers are installed before the first worktree exists and
 * uninstalled once the run settles, either way it settled.
 */
export async function runWithWorktrees<M, R>(
  phase: (repoDir: string, worktrees: WorktreeInfo[], signal: AbortSignal) => Promise<M>,
  buildResult: (measurement: M, cleanup: CleanupResult) => R,
): Promise<R> {
  const repoDir = process.cwd();
  const worktrees: WorktreeInfo[] = [];
  const run = new AbortController();

  const uninstallTerminationCleanup = installTerminationCleanup(() => {
    // Kill first: the bench group is detached, so gymrat's own Ctrl-C never
    // reaches it and it would outlive the process. SIGKILL delivery is not
    // awaited — removal may well race a still-dying process — but unlinking a
    // directory out from under a live cwd is legal on POSIX, so the sweep
    // succeeds either way.
    run.abort();
    cleanupWorktrees(worktrees, repoDir);
  });

  const runPhase = async (): Promise<R> => {
    let measurement: M;

    try {
      measurement = await phase(repoDir, worktrees, run.signal);
    } catch (error) {
      const cleanup = cleanupWorktrees(worktrees, repoDir);
      throw withCleanupFailures(error, cleanup);
    }

    const cleanup = cleanupWorktrees(worktrees, repoDir);
    return buildResult(measurement, cleanup);
  };

  return runPhase().finally(uninstallTerminationCleanup);
}

File: src/gymrat/sampling.py
"""Sequential sample collection for a benchmark comparison.

Runs a bench command a fixed number of times against each target, parsing every
run's output into a metric record. Runs are strictly sequential and interleaved
by round: every target is sampled once before any target is sampled again, so a
transient slowdown on the machine spreads across both sides rather than skewing
one. An optional prepare command runs once per target before sampling begins.

A single policy site turns a failed command (non-zero exit or timeout) into a
:class:`~gymrat.errors.CommandError`, so a failure anywhere stops the schedule
with the same formatted diagnosis.
"""

import asyncio
import math
from collections.abc import Awaitable, Callable, Sequence
from dataclasses import dataclass
from pathlib import Path
from typing import Literal

from gymrat.adapters.types import Adapter, WarnSink
from gymrat.config import KindEntry, MetricEntry, resolve_metric_meta
from gymrat.errors import CommandError, GymratError, hint_of
from gymrat.exec import ExecOptions, ExecResult, ExecTimeoutError, kill_live_process_groups
from gymrat.exec import (
    exec as exec,  # noqa: A004, PLC0414 -- names the subprocess executor `exec`
)
from gymrat.model import ResolvedMetricMeta
from gymrat.progress_events import (
    PassFinished,
    PassStarted,
    PrepareFinished,
    PrepareStarted,
    ProgressCallback,
    default_clock,
    emit_progress,
)
from gymrat.report.text import format_cleanup_failures
from gymrat.signals import install_termination_cleanup
from gymrat.stats.descriptive import compute_half_range, compute_median
from gymrat.targets import (
    CleanupResult,
    RefTarget,
    Target,
    WorktreeInfo,
    cleanup_worktrees,
    materialize_worktree,
    plan_worktree,
)


@dataclass(frozen=True, slots=True)
class TargetSpec:
    """One target a comparison or measurement names, before resolution.

    Attributes:
        label: An explicit display label, or ``None`` to derive one from the
            resolved target (a ref's name or a directory's basename).
        target: A git ref (resolved to a throwaway worktree) or a filesystem
            directory path (benched in place).
    """

    label: str | None
    target: str


@dataclass(frozen=True, slots=True)
class TargetContext:
    """A target paired with where and how it is run.

    Attributes:
        target: The thing being benchmarked.
        dir: The directory the command runs in.
        label: The target's display label.
        position: Which side of a comparison the target occupies, or ``None``
            when the run is not a two-sided comparison.
    """

    target: Target
    dir: str
    label: str
    position: Literal["old", "new"] | None = None


@dataclass(frozen=True, slots=True)
class TargetSamples:
    """Every metric record collected for one target, with its context.

    Attributes:
        ctx: The context the samples were collected under.
        samples: One metric record per successful bench run, in round order.
    """

    ctx: TargetContext
    samples: list[dict[str, float]]


@dataclass(frozen=True, slots=True)
class SamplingOptions:
    """Inputs governing a sampling run.

    Attributes:
        bench: The command run once per target per round.
        prepare: A command run once per target before sampling, or ``None``.
        samples: The number of rounds.
        timeout_seconds: Per-command wall-clock budget, in seconds.
        on_progress: Invoked with each progress event, or ``None`` for silence.
        warn: Where an adapter sends complaints about output it could not read,
            or ``None`` to use the adapter's own default.
        clock: A source of monotonic millisecond timestamps for event stamping.
            Defaults to :func:`~gymrat.progress_events.default_clock`.
    """

    bench: str
    prepare: str | None
    samples: int
    timeout_seconds: float
    on_progress: ProgressCallback | None = None
    warn: WarnSink | None = None
    clock: Callable[[], float] = default_clock


@dataclass(frozen=True, slots=True, kw_only=True)
class RunOptions:
    """The run settings a comparison and a measurement both take.

    Beyond the sampling fields :class:`SamplingOptions` reads, this adds the
    three inputs a caller needs to turn raw samples into a report: which adapter
    parses the bench output, and the per-metric and per-kind config overrides
    that settle each metric's metadata.

    Attributes:
        bench: The command run once per target per round.
        prepare: A command run once per target before sampling, or ``None``.
        adapter: Which output format ``bench`` writes, by adapter name.
        samples: The number of rounds.
        timeout_seconds: Per-command wall-clock budget, in seconds.
        config_metrics: Per-metric overrides from config, or ``None``.
        config_kinds: Per-kind overrides from config, or ``None``.
        on_progress: Invoked with each progress event, or ``None`` for silence.
        warn: Where an adapter sends complaints about unreadable output, or
            ``None`` to use the adapter's own default.
    """

    bench: str
    prepare: str | None
    adapter: str
    samples: int
    timeout_seconds: float
    config_metrics: dict[str, MetricEntry] | None
    config_kinds: dict[str, KindEntry] | None
    on_progress: ProgressCallback | None = None
    warn: WarnSink | None = None


@dataclass(frozen=True, slots=True)
class MetricStats:
    """A metric's central value and relative spread.

    Attributes:
        median: The metric's median, or ``None`` when there were no values.
        spread: The half-range as a percentage of the median's magnitude, or
            ``None`` when it is undefined (fewer than two values, a zero median,
            or a non-finite ratio).
    """

    median: float | None
    spread: float | None


_REF_HINT = (
    "the worktree only contains files tracked at this ref; "
    "untracked, gitignored, or not-yet-committed files are absent"
)
_LABEL_WIDTH = 11
_MIN_SPREAD_SAMPLES = 2


async def collect_samples(
    adapter: Adapter,
    targets: Sequence[TargetContext],
    options: SamplingOptions,
    abort: asyncio.Event,
) -> list[TargetSamples]:
    """Run the prepare and bench schedule and collect each target's samples.

    Prepare (when set) runs once per target in order before any bench run. Then
    for each round, bench runs once per target in order, so round ``n+1`` never
    starts before round ``n`` has run every target. Each successful bench run
    contributes one parsed metric record to its target.

    Args:
        adapter: Parses a bench run's stdout into a metric record.
        targets: The targets to sample, in the order they are run and returned.
        options: The bench and prepare commands, round count, timeout, and hooks.
        abort: An event whose being set kills the in-flight command; passed
            through to the command runner.

    Returns:
        One :class:`TargetSamples` per target, in target order.

    Raises:
        CommandError: A prepare or bench command timed out or exited non-zero.
    """
    timeout_ms = int(options.timeout_seconds * 1000)
    collected: list[list[dict[str, float]]] = [[] for _ in targets]
    target_count = len(targets)

    if options.prepare is not None:
        for ctx in targets:
            emit_progress(
                options.on_progress, PrepareStarted(label=ctx.label, at_ms=options.clock())
            )
            await _run_command("prepare", None, options.prepare, ctx, timeout_ms, abort)
            emit_progress(
                options.on_progress, PrepareFinished(label=ctx.label, at_ms=options.clock())
            )

    for round_index in range(options.samples):
        for target_index, ctx in enumerate(targets):
            emit_progress(
                options.on_progress,
                PassStarted(
                    round=round_index + 1,
                    total_rounds=options.samples,
                    target_count=target_count,
                    label=ctx.label,
                    at_ms=options.clock(),
                ),
            )
            stdout = await _run_command(
                "bench", round_index + 1, options.bench, ctx, timeout_ms, abort
            )
            emit_progress(
                options.on_progress,
                PassFinished(
                    round=round_index + 1,
                    total_rounds=options.samples,
                    target_count=target_count,
                    label=ctx.label,
                    at_ms=options.clock(),
                ),
            )
            collected[target_index].append(_parse(adapter, stdout, options.warn))

    return [
        TargetSamples(ctx=ctx, samples=samples)
        for ctx, samples in zip(targets, collected, strict=True)
    ]


async def _run_command(  # noqa: PLR0913, PLR0917 -- one parameter per command-execution axis
    phase: str,
    sample_index: int | None,
    command: str,
    ctx: TargetContext,
    timeout_ms: int,
    abort: asyncio.Event,
) -> str:
    """Run one command and return its stdout, or raise on failure."""
    result = await exec(command, ExecOptions(cwd=ctx.dir, timeout_ms=timeout_ms, abort=abort))
    if isinstance(result, ExecTimeoutError) or result.exit_code != 0:
        raise _to_command_error(phase, sample_index, command, ctx, result, timeout_ms)
    return result.stdout


def _to_command_error(  # noqa: PLR0913, PLR0917 -- one field per failure axis
    phase: str,
    sample_index: int | None,
    command: str,
    ctx: TargetContext,
    result: ExecResult | ExecTimeoutError,
    request_timeout_ms: int,
) -> CommandError:
    """Map a command failure to a target-specific :class:`CommandError`.

    A ref target contributes ``ref`` and ``worktree`` location lines plus the
    hint that the worktree only holds tracked files; a plain directory
    contributes a single ``dir`` line and no hint.
    """
    timed_out = isinstance(result, ExecTimeoutError)
    if timed_out:
        timeout_ms = result.timeout_ms
        exit_code: int | None = None
    else:
        timeout_ms = request_timeout_ms
        exit_code = result.exit_code

    target = ctx.target
    if isinstance(target, RefTarget):
        location = [_field("ref", target.ref), _field("worktree", ctx.dir)]
        hint = _REF_HINT
    else:
        location = [_field("dir", ctx.dir)]
        hint = None

    position = f"{ctx.position}, " if ctx.position is not None else ""
    sample = f", sample {sample_index}" if sample_index is not None else ""
    outcome_label = "timed out" if timed_out else "failed"
    header = f'{phase} command {outcome_label} ({position}"{ctx.label}"{sample})'

    lines = [header, *location, _field("command", command)]
    if timed_out:
        lines.append(_field("timeout", f"{timeout_ms}ms"))
    else:
        lines.append(_field("exit code", exit_code))
    lines.extend(
        _captured_output(result.stdout, result.stdout_bytes, result.stderr, result.stderr_bytes)
    )

    return CommandError("\n".join(lines), hint=hint)


def _field(label: str, value: object) -> str:
    """Format an indented, column-aligned ``label: value`` detail line."""
    return f"  {(label + ':').ljust(_LABEL_WIDTH)}{value}"


def _captured_output(stdout: str, stdout_bytes: int, stderr: str, stderr_bytes: int) -> list[str]:
    """Render whatever the failed command wrote to its output streams.

    A lone non-empty stream is emitted bare unless its captured text was
    truncated, in which case it — like every stream when both are present —
    becomes a labeled entry annotated with the true byte total.
    """
    streams = [
        ("stderr", stderr, stderr_bytes),
        ("stdout", stdout, stdout_bytes),
    ]
    present = [(label, text, total) for label, text, total in streams if text]

    if len(present) == 1:
        label, text, total = present[0]
        if not _is_truncated(text, total):
            return [text]
        return _labeled(label, text, total)

    return [line for entry in present for line in _labeled(*entry)]


def _is_truncated(text: str, total_bytes: int) -> bool:
    """Whether ``total_bytes`` exceeds what survived capture in ``text``."""
    return total_bytes > len(text.encode())


def _labeled(label: str, text: str, total: int) -> list[str]:
    """Build a labeled stream entry, flagging truncation against the byte total."""
    suffix = f" (truncated, {total} bytes total)" if _is_truncated(text, total) else ""
    return [f"--- {label}{suffix} ---", text]


def _parse(adapter: Adapter, stdout: str, warn: WarnSink | None) -> dict[str, float]:
    """Parse a bench run's stdout, routing warnings through ``warn`` when given."""
    if warn is None:
        return adapter.parse(stdout)
    return adapter.parse(stdout, warn)


def compute_metric_stats(values: Sequence[float]) -> MetricStats:
    """Summarize a metric's samples as a median and relative spread.

    Args:
        values: The metric's sampled values.

    Returns:
        The median and its half-range as a percentage of ``abs(median)``. The
        spread is absent when there are fewer than two values, the median is
        zero, or the ratio is non-finite.
    """
    if not values:
        return MetricStats(median=None, spread=None)

    median = compute_median(values)
    if len(values) < _MIN_SPREAD_SAMPLES or median == 0:
        return MetricStats(median=median, spread=None)

    ratio = compute_half_range(values) / abs(median) * 100
    if not math.isfinite(ratio):
        return MetricStats(median=median, spread=None)
    return MetricStats(median=median, spread=ratio)


def own_values(samples: Sequence[dict[str, float]], name: str) -> list[float]:
    """Collect the values a side reported for ``name``, skipping rounds without it.

    Args:
        samples: One metric record per round.
        name: The metric to extract.

    Returns:
        The reported values for ``name``, in round order.
    """
    return [record[name] for record in samples if name in record]


def paired_or_own_values(
    paired: Sequence[float],
    samples: Sequence[dict[str, float]],
    name: str,
) -> list[float]:
    """Prefer already-paired values, falling back to a side's own values.

    Args:
        paired: Values paired across sides; used as-is when non-empty.
        samples: One metric record per round, used only for the fallback.
        name: The metric to extract when falling back.

    Returns:
        ``paired`` when it holds any values, otherwise ``own_values(samples, name)``.
    """
    if paired:
        return list(paired)
    return own_values(samples, name)


def resolve_metric_meta_from_samples(
    sample_sets: Sequence[list[dict[str, float]]],
    config_metrics: dict[str, MetricEntry] | None,
    adapter: Adapter,
    config_kinds: dict[str, KindEntry] | None = None,
) -> dict[str, ResolvedMetricMeta]:
    """Collect every metric name across the sample sets and resolve its metadata.

    The union of names is taken in first-appearance order across the flattened
    samples, so the resolved metadata — and every report drawn from it — reads in
    the order the run first reported each metric.

    Args:
        sample_sets: One list of per-round metric records per target.
        config_metrics: Per-metric overrides from config, or ``None``.
        adapter: The adapter whose defaults seed each metric's metadata.
        config_kinds: Per-kind overrides from config, or ``None``.

    Returns:
        The resolved metadata for each metric, keyed by metric name.

    Raises:
        GymratError: No sample set reported any metric. Adapters reject empty
            output themselves, so this guards the otherwise-unreachable case.
    """
    names = dict.fromkeys(name for samples in sample_sets for sample in samples for name in sample)
    if not names:
        message = "No metrics found in benchmark output"
        raise GymratError(message)

    return resolve_metric_meta(list(names), config_metrics, adapter, config_kinds)


def resolve_dir(target: Target, repo_dir: str, worktrees: list[WorktreeInfo]) -> str:
    """The directory a target runs in, materializing a worktree for a ref.

    A ref is benchmarked from its own worktree. The planned worktree is appended
    to ``worktrees`` before ``git worktree add`` runs, so a caller sweeping the
    registry on termination can remove a directory a killed add left behind.

    Args:
        target: The target to locate.
        repo_dir: The repository the worktree is added from.
        worktrees: The live registry of claimed worktrees, appended to in place.

    Returns:
        The directory the benchmark runs in.
    """
    if isinstance(target, RefTarget):
        worktree = plan_worktree(target)
        worktrees.append(worktree)
        materialize_worktree(worktree, repo_dir)
        return worktree.dir
    return target.dir


def resolve_label(explicit: str | None, target: Target) -> str:
    """The display label for a target.

    Args:
        explicit: A caller-supplied label, or ``None`` to derive one.
        target: The target a label is derived from when ``explicit`` is ``None``.

    Returns:
        ``explicit`` when given, else a ref's name, else an in-place target's
        directory basename.
    """
    if explicit is not None:
        return explicit
    if isinstance(target, RefTarget):
        return target.ref
    return Path(target.dir).name


async def run_with_worktrees[M, R](
    phase: Callable[[str, list[WorktreeInfo], asyncio.Event], Awaitable[M]],
    build_result: Callable[[M, CleanupResult], R],
) -> R:
    """Run a phase that may claim worktrees, sweeping them on every exit path.

    A termination cleanup is installed before any worktree exists, so a signal
    arriving mid-run still sweeps whatever was claimed; that cleanup aborts the
    run and sweeps. The normal path sweeps exactly once whether the phase returns
    or raises. When the sweep leaves worktrees behind, the phase's error is
    re-raised wrapped with the cleanup diagnostics.

    Args:
        phase: The work to run. It receives the repository directory, the
            registry it appends claimed worktrees to, and an abort event a
            termination signal sets.
        build_result: Combines the phase's measurement with the cleanup outcome
            into the return value.

    Returns:
        The value ``build_result`` produced from the measurement and cleanup.
    """
    repo_dir = str(Path.cwd())
    worktrees: list[WorktreeInfo] = []
    abort = asyncio.Event()

    def terminate() -> None:
        abort.set()
        # Kill any live bench group synchronously: on the signal path the loop
        # may not resume to process the abort before the sweep runs, so the
        # child must be dead before cleanup_worktrees touches the worktrees.
        kill_live_process_groups()
        cleanup_worktrees(worktrees, repo_dir)

    uninstall = install_termination_cleanup(terminate)
    try:
        try:
            measurement = await phase(repo_dir, worktrees, abort)
        except Exception as error:
            cleanup = cleanup_worktrees(worktrees, repo_dir)
            wrapped = _with_cleanup_failures(error, cleanup)
            if wrapped is error:
                raise
            raise wrapped from error
        cleanup = cleanup_worktrees(worktrees, repo_dir)
        return build_result(measurement, cleanup)
    finally:
        uninstall()


def _with_cleanup_failures(error: Exception, cleanup: CleanupResult) -> Exception:
    """Fold cleanup diagnostics into ``error``, preserving its subclass and hint.

    Returns ``error`` unchanged when the sweep was clean. Otherwise returns a new
    exception of the same type carrying the original message plus the cleanup
    diagnostics, with ``error`` chained as its cause.

    ``type(error)(...)`` reconstructs the exact subclass — a
    :class:`~gymrat.errors.CommandError` stays a ``CommandError`` and an
    ``AdapterError`` stays an ``AdapterError`` — because every
    :class:`~gymrat.errors.GymratError` shares the ``(message, *, hint)``
    signature. A non-gymrat error has no hint and becomes a plain ``Exception``.

    Args:
        error: The error the phase raised.
        cleanup: The outcome of the worktree sweep.

    Returns:
        ``error`` when the sweep left nothing behind, else a same-typed
        replacement whose message appends the cleanup diagnostics.
    """
    details = format_cleanup_failures(cleanup.failures, cleanup.prune_error)
    if not details:
        return error

    combined = "\n".join([str(error), "", "cleanup did not finish:", *details])
    if isinstance(error, GymratError):
        wrapped: Exception = type(error)(combined, hint=hint_of(error))
    else:
        wrapped = Exception(combined)
    return wrapped
